src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewarePipelineService.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Collections.Concurrent;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    /// <summary>
+    /// Builds a middleware pipeline after receiving the pipeline from a pipeline provider
+    /// </summary>
+    public class MiddlewarePipelineBuilderService
+    {
+        private readonly ConcurrentDictionary<Type, RequestDelegate> _pipelinesCache
+            = new ConcurrentDictionary<Type, RequestDelegate>();
+
+        public IApplicationBuilder ApplicationBuilder { get; set; }
+
+        public RequestDelegate GetPipeline(Type type)
+        {
+            // Build the pipeline only once. This is similar to how middlewares are used where they are constructed
+            // only once.
+
+            RequestDelegate requestDelegate;
+            if (!_pipelinesCache.TryGetValue(type, out requestDelegate))
```

Comment:
This isn't thread safe, you could end up building the same pipeline multiple times in parallel. That's not something middleware normally experience in Startup and it may cause issues.